File: packages/go-dal/storage/s3.go
package storage

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"path"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/aws/aws-sdk-go-v2/service/s3/types"
	"github.com/penguintechinc/penguin-libs/packages/go-dal"
)

// s3ObjectClient is the subset of *s3.Client operations used by S3Store.
type s3ObjectClient interface {
	PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error)
	GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error)
	DeleteObject(ctx context.Context, params *s3.DeleteObjectInput, optFns ...func(*s3.Options)) (*s3.DeleteObjectOutput, error)
	HeadObject(ctx context.Context, params *s3.HeadObjectInput, optFns ...func(*s3.Options)) (*s3.HeadObjectOutput, error)
	NewListObjectsV2Paginator(params *s3.ListObjectsV2Input) s3PageIterator
}

// s3PageIterator abstracts the ListObjectsV2Paginator for testability.
type s3PageIterator interface {
	HasMorePages() bool
	NextPage(ctx context.Context, optFns ...func(*s3.Options)) (*s3.ListObjectsV2Output, error)
}

// s3PresignClient is the subset of *s3.PresignClient operations used by S3Store.
type s3PresignClient interface {
	PresignGetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.PresignOptions)) (*s3PresignResult, error)
}

// s3PresignResult holds the URL returned by presigning.
type s3PresignResult struct {
	URL string
}

// realS3Client wraps *s3.Client to implement s3ObjectClient.
type realS3Client struct {
	inner *s3.Client
}

func (r *realS3Client) PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error) {
	return r.inner.PutObject(ctx, params, optFns...)
}

func (r *realS3Client) GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
	return r.inner.GetObject(ctx, params, optFns...)
}

func (r *realS3Client) DeleteObject(ctx context.Context, params *s3.DeleteObjectInput, optFns ...func(*s3.Options)) (*s3.DeleteObjectOutput, error) {
	return r.inner.DeleteObject(ctx, params, optFns...)
}

func (r *realS3Client) HeadObject(ctx context.Context, params *s3.HeadObjectInput, optFns ...func(*s3.Options)) (*s3.HeadObjectOutput, error) {
	return r.inner.HeadObject(ctx, params, optFns...)
}

func (r *realS3Client) NewListObjectsV2Paginator(params *s3.ListObjectsV2Input) s3PageIterator {
	return s3.NewListObjectsV2Paginator(r.inner, params)
}

// realS3PresignClient wraps *s3.PresignClient to implement s3PresignClient.
type realS3PresignClient struct {
	inner *s3.PresignClient
}

func (r *realS3PresignClient) PresignGetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.PresignOptions)) (*s3PresignResult, error) {
	result, err := r.inner.PresignGetObject(ctx, params, optFns...)
	if err != nil {
		return nil, err
	}
	return &s3PresignResult{URL: result.URL}, nil
}

// S3Config configures an S3 storage backend.
type S3Config struct {
	Bucket      string
	Region      string
	AccessKey   string
	SecretKey   string
	EndpointURL string
	Prefix      string
}

// S3Store implements dal.StorageStore for AWS S3.
type S3Store struct {
	cfg     S3Config
	client  s3ObjectClient
	presign s3PresignClient
}

// NewS3Store creates a new S3 storage backend.
func NewS3Store(ctx context.Context, cfg S3Config) (*S3Store, error) {
	if cfg.Bucket == "" {
		return nil, fmt.Errorf("go-dal: s3: %w: bucket required", dal.ErrInvalidConfiguration)
	}
	if cfg.Region == "" {
		return nil, fmt.Errorf("go-dal: s3: %w: region required", dal.ErrInvalidConfiguration)
	}

	awsOpts := []func(*config.LoadOptions) error{
		config.WithRegion(cfg.Region),
	}

	if cfg.AccessKey != "" && cfg.SecretKey != "" {
		awsOpts = append(awsOpts, config.WithCredentialsProvider(credentials.NewStaticCredentialsProvider(
			cfg.AccessKey, cfg.SecretKey, "",
		)))
	}

	awsCfg, err := config.LoadDefaultConfig(ctx, awsOpts...)
	if err != nil {
		return nil, fmt.Errorf("go-dal: s3: load config: %w", err)
	}

	rawClient := s3.NewFromConfig(awsCfg)
	rawPresign := s3.NewPresignClient(rawClient)

	store := &S3Store{
		cfg:     cfg,
		client:  &realS3Client{inner: rawClient},
		presign: &realS3PresignClient{inner: rawPresign},
	}

	return store, nil
}

// NewS3StoreWithClient creates an S3Store using injected clients (for testing).
func NewS3StoreWithClient(client s3ObjectClient, presign s3PresignClient, cfg S3Config) *S3Store {
	return &S3Store{
		cfg:     cfg,
		client:  client,
		presign: presign,
	}
}

// Put writes data to S3.
func (s *S3Store) Put(ctx context.Context, key string, data []byte, opts ...dal.PutOption) error {
	po := &dal.PutOptions{}
	for _, opt := range opts {
		opt(po)
	}

	fullKey := path.Join(s.cfg.Prefix, key)
	input := &s3.PutObjectInput{
		Bucket: aws.String(s.cfg.Bucket),
		Key:    aws.String(fullKey),
		Body:   bytes.NewReader(data),
	}

	if po.ContentType != "" {
		input.ContentType = aws.String(po.ContentType)
	}
	if po.CacheControl != "" {
		input.CacheControl = aws.String(po.CacheControl)
	}
	if len(po.Metadata) > 0 {
		input.Metadata = po.Metadata
	}

	_, err := s.client.PutObject(ctx, input)
	if err != nil {
		return fmt.Errorf("go-dal: s3: put: %w", err)
	}

	return nil
}

// Get retrieves data from S3.
func (s *S3Store) Get(ctx context.Context, key string) ([]byte, error) {
	fullKey := path.Join(s.cfg.Prefix, key)
	input := &s3.GetObjectInput{
		Bucket: aws.String(s.cfg.Bucket),
		Key:    aws.String(fullKey),
	}

	output, err := s.client.GetObject(ctx, input)
	if err != nil {
		var nsk *types.NoSuchKey
		if _, ok := err.(*types.NoSuchKey); ok || nsk != nil {
			return nil, fmt.Errorf("go-dal: s3: get: %w", dal.ErrNotFound)
		}
		return nil, fmt.Errorf("go-dal: s3: get: %w", err)
	}
	defer output.Body.Close()

	data, err := io.ReadAll(output.Body)
	if err != nil {
		return nil, fmt.Errorf("go-dal: s3: get read: %w", err)
	}

	return data, nil
}

// Delete removes an object from S3.
func (s *S3Store) Delete(ctx context.Context, key string) error {
	fullKey := path.Join(s.cfg.Prefix, key)
	input := &s3.DeleteObjectInput{
		Bucket: aws.String(s.cfg.Bucket),
		Key:    aws.String(fullKey),
	}

	_, err := s.client.DeleteObject(ctx, input)
	if err != nil {
		return fmt.Errorf("go-dal: s3: delete: %w", err)
	}

	return nil
}

// Exists checks if an object exists in S3.
func (s *S3Store) Exists(ctx context.Context, key string) (bool, error) {
	fullKey := path.Join(s.cfg.Prefix, key)
	input := &s3.HeadObjectInput{
		Bucket: aws.String(s.cfg.Bucket),
		Key:    aws.String(fullKey),
	}

	_, err := s.client.HeadObject(ctx, input)
	if err != nil {
		var nsk *types.NoSuchKey
		if _, ok := err.(*types.NoSuchKey); ok || nsk != nil {
			return false, nil
		}
		return false, fmt.Errorf("go-dal: s3: exists: %w", err)
	}

	return true, nil
}

// List returns all objects with a given prefix.
func (s *S3Store) List(ctx context.Context, prefix string) ([]string, error) {
	fullPrefix := path.Join(s.cfg.Prefix, prefix)
	var keys []string
	paginator := s.client.NewListObjectsV2Paginator(&s3.ListObjectsV2Input{
		Bucket: aws.String(s.cfg.Bucket),
		Prefix: aws.String(fullPrefix),
	})

	for paginator.HasMorePages() {
		output, err := paginator.NextPage(ctx)
		if err != nil {
			return nil, fmt.Errorf("go-dal: s3: list: %w", err)
		}

		for _, obj := range output.Contents {
			keys = append(keys, *obj.Key)
		}
	}

	return keys, nil
}

// GetURL generates a presigned URL for temporary access.
func (s *S3Store) GetURL(ctx context.Context, key string, expiresIn time.Duration) (string, error) {
	fullKey := path.Join(s.cfg.Prefix, key)
	input := &s3.GetObjectInput{
		Bucket: aws.String(s.cfg.Bucket),
		Key:    aws.String(fullKey),
	}

	presignOptFn := func(opts *s3.PresignOptions) {
		opts.Expires = expiresIn
	}

	result, err := s.presign.PresignGetObject(ctx, input, presignOptFn)
	if err != nil {
		return "", fmt.Errorf("go-dal: s3: presign: %w", err)
	}

	return result.URL, nil
}

// Close closes the S3 client (no-op, client is stateless).
func (s *S3Store) Close() error {
	return nil
}

File: packages/python-dal/src/penguin_dal/storage/s3.py
"""S3-compatible object storage backend."""
from __future__ import annotations

import asyncio
from dataclasses import dataclass

from penguin_dal.protocols import PutOptions, StorageStore


@dataclass(slots=True)
class S3Config:
    """Configuration for S3-compatible storage."""

    bucket: str
    region: str = "us-east-1"
    access_key: str | None = None
    secret_key: str | None = None
    endpoint_url: str | None = None
    prefix: str = ""


class S3Store:
    """Synchronous S3-compatible object store using boto3."""

    def __init__(self, config: S3Config) -> None:
        """Initialize S3 store with configuration."""
        try:
            import boto3
        except ImportError as e:
            raise ImportError(
                "boto3 is required for S3 storage. Install with: pip install penguin-dal[s3]"
            ) from e

        self._config = config
        self._boto3 = boto3

        session_kwargs: dict[str, str | None] = {}
        if config.access_key:
            session_kwargs["aws_access_key_id"] = config.access_key
        if config.secret_key:
            session_kwargs["aws_secret_access_key"] = config.secret_key

        session = boto3.Session(**{k: v for k, v in session_kwargs.items() if v})
        client_kwargs: dict[str, str] = {"region_name": config.region}
        if config.endpoint_url:
            client_kwargs["endpoint_url"] = config.endpoint_url

        self._client = session.client("s3", **client_kwargs)

    def _make_key(self, key: str) -> str:
        """Build full S3 key with prefix."""
        if self._config.prefix:
            return f"{self._config.prefix}/{key}"
        return key

    def _strip_prefix(self, key: str) -> str:
        """Remove prefix from key for listing results."""
        if self._config.prefix and key.startswith(f"{self._config.prefix}/"):
            return key[len(self._config.prefix) + 1 :]
        return key

    def put(
        self, key: str, data: bytes, opts: PutOptions | None = None
    ) -> None:
        """Store object at key."""
        opts = opts or PutOptions()
        s3_key = self._make_key(key)

        put_kwargs: dict[str, str | dict[str, str]] = {
            "ContentType": opts.content_type,
        }

        if opts.metadata:
            put_kwargs["Metadata"] = opts.metadata

        if opts.cache_control:
            put_kwargs["CacheControl"] = opts.cache_control

        self._client.put_object(
            Bucket=self._config.bucket,
            Key=s3_key,
            Body=data,
            **put_kwargs,
        )

    def get(self, key: str) -> bytes:
        """Retrieve object at key. Raises KeyError if not found."""
        s3_key = self._make_key(key)
        try:
            response = self._client.get_object(
                Bucket=self._config.bucket,
                Key=s3_key,
            )
            return response["Body"].read()
        except self._client.exceptions.NoSuchKey:
            raise KeyError(key)
        except Exception as e:
            if "NoSuchKey" in str(e) or "404" in str(e):
                raise KeyError(key)
            raise

    def delete(self, key: str) -> None:
        """Delete object at key."""
        s3_key = self._make_key(key)
        self._client.delete_object(
            Bucket=self._config.bucket,
            Key=s3_key,
        )

    def exists(self, key: str) -> bool:
        """Check if object exists at key."""
        s3_key = self._make_key(key)
        try:
            self._client.head_object(
                Bucket=self._config.bucket,
                Key=s3_key,
            )
            return True
        except self._client.exceptions.NoSuchKey:
            return False
        except Exception as e:
            if "NoSuchKey" in str(e) or "404" in str(e):
                return False
            raise

    def list(self, prefix: str = "") -> list[str]:
        """List all keys with given prefix."""
        s3_prefix = self._make_key(prefix)
        keys: list[str] = []

        paginator = self._client.get_paginator("list_objects_v2")
        pages = paginator.paginate(
            Bucket=self._config.bucket,
            Prefix=s3_prefix,
        )

        for page in pages:
            if "Contents" in page:
                for obj in page["Contents"]:
                    full_key = obj["Key"]
                    stripped_key = self._strip_prefix(full_key)

                    if prefix:
                        if prefix in full_key:
                            relative = full_key[len(s3_prefix) + 1 :]
                            if relative:
                                keys.append(relative)
                    else:
                        if stripped_key:
                            keys.append(stripped_key)

        return sorted(keys)

    def get_url(self, key: str, expires_in: int = 3600) -> str:
        """Get a presigned URL to access object at key."""
        s3_key = self._make_key(key)
        return self._client.generate_presigned_url(
            "get_object",
            Params={"Bucket": self._config.bucket, "Key": s3_key},
            ExpiresIn=expires_in,
        )


class AsyncS3Store:
    """Asynchronous S3 store wrapping S3Store via asyncio.to_thread."""

    def __init__(self, config: S3Config) -> None:
        """Initialize async S3 store."""
        self._sync_store = S3Store(config)

    async def put(
        self, key: str, data: bytes, opts: PutOptions | None = None
    ) -> None:
        """Store object at key."""
        await asyncio.to_thread(self._sync_store.put, key, data, opts)

    async def get(self, key: str) -> bytes:
        """Retrieve object at key. Raises KeyError if not found."""
        return await asyncio.to_thread(self._sync_store.get, key)

    async def delete(self, key: str) -> None:
        """Delete object at key."""
        await asyncio.to_thread(self._sync_store.delete, key)

    async def exists(self, key: str) -> bool:
        """Check if object exists at key."""
        return await asyncio.to_thread(self._sync_store.exists, key)

    async def list(self, prefix: str = "") -> list[str]:
        """List all keys with given prefix."""
        return await asyncio.to_thread(self._sync_store.list, prefix)

    async def get_url(self, key: str, expires_in: int = 3600) -> str:
        """Get a presigned URL to access object at key."""
        return await asyncio.to_thread(self._sync_store.get_url, key, expires_in)
